src/transform/src/normalize_lets.rs

```diff
@@ -0,0 +1,361 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Normalize the structure of `Let` operators in expressions.
+//!
+//! Let operators are hoisted to the root of the expression, and given
+//! consecutive identifiers from zero. Each bound `value` is `Let`-free,
+//! and both is more complex than a `Get` or a `Constant` and occurs at
+//! least twice in the `body` of its binding.
+//!
+//! No effort is yet made to choose among the topological ordors of the
+//! bindings in some canonical way; this could be future work if we are
+//! excited amount it (I think we don't expect CSE to discover matches
+//! involving `Let` operator as they are all at the roots).
+//!
+//! The transform may remove some `Let` and `Get` operators, and does not
+//! introduce any new operators. It is idempotent.
+//!
+//! The module also publishes the function `renumber_bindings` which can
+//! be used to renumber bindings in an expression starting from a provided
+//! `IdGen`, which is used to prepare distinct expressions for inlining.
+use std::collections::BTreeMap;
+
+use mz_expr::RECURSION_LIMIT;
+use mz_expr::{Id, LocalId, MirRelationExpr};
+use mz_ore::id_gen::IdGen;
+use mz_ore::stack::RecursionGuard;
+use mz_repr::RelationType;
+
+use crate::TransformArgs;
+
+/// Install replace certain `Get` operators with their `Let` value.
+#[derive(Debug)]
+pub struct NormalizeLets {
+    /// If `true`, inline MFPs around a Get.
+    ///
+    /// We want this value to be true for the NormalizeLets call that comes right
+    /// before [crate::join_implementation::JoinImplementation] runs because
+    /// [crate::join_implementation::JoinImplementation] cannot lift MFPs
+    /// through a Let.
+    ///
+    /// Generally, though, we prefer to be more conservative in our inlining in
+    /// order to be able to better detect CSEs.
+    pub inline_mfp: bool,
+}
+
+impl NormalizeLets {
+    /// Construct a new [`NormalizeLets`] instance with the given `inline_mfp`.
+    pub fn new(inline_mfp: bool) -> NormalizeLets {
+        NormalizeLets { inline_mfp }
+    }
+}
+
+impl crate::Transform for NormalizeLets {
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "normalize_lets")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _args: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        let result = self.transform_without_trace(relation);
+        mz_repr::explain_new::trace_plan(&*relation);
+        result
+    }
+}
+
+impl NormalizeLets {
+    /// Performs the `NormalizeLets` transformation without tracing the result.
+    pub fn transform_without_trace(
+        &self,
+        relation: &mut MirRelationExpr,
+    ) -> Result<(), crate::TransformError> {
+        self.action(relation)?;
+        relation.typ();
```

Comment:
What is the intent of this call?  This seems to be the only place where `typ` is used for a side effect alone, and it's not clear to me what side effect that could be.

Reply:
It's vestigial. Was there to debug a thing, and somehow it missed my best efforts to find and remove all of these. :)

Reply:
The side effect, fwiw, is validating the type of the expression. Some of our code and some of this PR interacted to leave an inconsistency in the type, and this was here to check if things were well typed at the end of the transform.